### PR TITLE
Fix gas tracking with scheduled sc calls

### PR DESCRIPTION
**Pull request by AdoAdoAdo**

### CODE REVIEW

File: process/block/postprocess/feeHandler.go [hunk 1]
```diff
@@ -33,11 +34,18 @@ func NewFeeAccumulator() (*feeHandler, error) {
...
 
 // CreateBlockStarted does the cleanup before creating a new block
-func (f *feeHandler) CreateBlockStarted() {
+func (f *feeHandler) CreateBlockStarted(gasAndFees scheduled.GasAndFees) {
 	f.mut.Lock()
 	f.mapHashFee = make(map[string]*feeData)
-	f.accumulatedFees = big.NewInt(0)
-	f.developerFees = big.NewInt(0)
+	f.accumulatedFees = big.NewInt(0).Set(gasAndFees.AccumulatedFees)
+	if f.accumulatedFees == nil {
```
Comment (raduchis) [inline process/block/postprocess/feeHandler.go:41]: is the check necessary? there is an assignment right above

File: process/block/postprocess/feeHandler.go [hunk 2]
```diff
@@ -33,11 +34,18 @@ func NewFeeAccumulator() (*feeHandler, error) {
...
+		f.accumulatedFees = big.NewInt(0)
+	}
+
+	f.developerFees = big.NewInt(0).Set(gasAndFees.DeveloperFees)
+	if f.developerFees == nil {
```
Comment (raduchis) [inline process/block/postprocess/feeHandler.go:46]: is the check necessary? there is an assignment right above

File: process/block/postprocess/feeHandler_test.go [hunk 3]
```diff
@@ -89,3 +92,13 @@ func TestFeeHandler_IsInterfaceNil(t *testing.T) {
...
+
+func getZeroGasAndFees() scheduled.GasAndFees {
+	return scheduled.GasAndFees{
+		AccumulatedFees: big.NewInt(0),
+		DeveloperFees:   big.NewInt(0),
+		GasProvided:     0,
+		GasPenalized:    0,
+		GasRefunded:     0,
+	}
+}
```
Comment (raduchis) [inline process/block/postprocess/feeHandler_test.go:104]: missing empty line at the end of file

File: process/block/preprocess/scheduledTxsExecution.go [hunk 4]
```diff
@@ -246,13 +255,21 @@ func (ste *scheduledTxsExecution) GetScheduledSCRs() map[block.Type][]data.Trans
...
+// GetScheduledGasAndFee returns the scheduled SC calls gas and fee
+func (ste *scheduledTxsExecution) GetScheduledGasAndFee() scheduled.GasAndFees {
+	ste.mutScheduledTxs.RLock()
+	defer ste.mutScheduledTxs.RUnlock()
+
+	return ste.gasAndFees
+}
+
+// SetScheduledRootHasSCRsAndGas sets the resulted scheduled root hash, SCRs and gas after the execution of scheduled transactions
+func (ste *scheduledTxsExecution) SetScheduledRootHasSCRsAndGas(rootHash []byte, mapSCRs map[block.Type][]data.TransactionHandler, gasAndFees scheduled.GasAndFees) {
```
Comment (raduchis) [inline process/block/preprocess/scheduledTxsExecution.go:267]: here is the actual rename needed SetScheduledRootHas**H**SCRsAndGas

File: process/block/preprocess/scheduledTxsExecution.go [hunk 5]
```diff
@@ -270,7 +287,9 @@ func (ste *scheduledTxsExecution) SetScheduledRootHashAndSCRs(rootHash []byte, m
 		numScheduledSCRs += len(scrs)
 	}
 
-	log.Debug("scheduledTxsExecution.SetScheduledRootHashAndSCRs", "num of scheduled scrs", numScheduledSCRs)
+	ste.gasAndFees = gasAndFees
+
+	log.Debug("scheduledTxsExecution.SetScheduledRootHasSCRsAndGas", "num of scheduled scrs", numScheduledSCRs)
```
Comment (raduchis) [inline process/block/preprocess/scheduledTxsExecution.go:292]: also here SetScheduledRootHas**H**SCRsAndGas

File: process/block/preprocess/scheduledTxsExecution.go [hunk 6]
```diff
@@ -316,85 +366,98 @@ func (ste *scheduledTxsExecution) GetScheduledRootHashForHeader(
...
 	mapScheduledSCRs := ste.GetScheduledSCRs()
 	ste.mutScheduledTxs.RLock()
+	gasAndFees := ste.gasAndFees
 	numScheduledTxs := len(ste.scheduledTxs)
 	ste.mutScheduledTxs.RUnlock()
 	log.Debug("scheduledTxsExecution.SaveStateIfNeeded", "num of scheduled txs", numScheduledTxs)
 
 	if numScheduledTxs > 0 {
-		ste.SaveState(headerHash, scheduledRootHash, mapScheduledSCRs)
+		ste.SaveState(headerHash, scheduledRootHash, mapScheduledSCRs, gasAndFees)
```
Comment (raduchis) [inline process/block/preprocess/scheduledTxsExecution.go:392]: no need to save gasAndFees? Could there be a case with no scheduledTxs but somehow the gasAndFees changed?

File: process/coordinator/process.go [hunk 7]
```diff
@@ -799,9 +802,17 @@ func (tc *transactionCoordinator) CreatePostProcessMiniBlocks() block.MiniBlockS
 	return miniBlocks
 }
 
+func (tc *transactionCoordinator) addGasForScheduled(gasAndFees scheduled.GasAndFees) {
+	tc.gasHandler.SetGasRefunded(gasAndFees.GasRefunded, []byte(prevScheduledTxs))
+	tc.gasHandler.SetGasConsumed(gasAndFees.GasProvided, []byte(prevScheduledTxs))
+	tc.gasHandler.SetGasPenalized(gasAndFees.GasRefunded, []byte(prevScheduledTxs))
```
Comment (raduchis) [inline process/coordinator/process.go:808]: gasAndFees.GasPenalized

An unit test for CreateBlockStarted could be added to validate the addGasForScheduled

File: process/sync/baseSync.go [hunk 8]
```diff
@@ -911,7 +920,14 @@ func (boot *baseBootstrap) restoreState(
...
 	if err != nil {
-		boot.scheduledTxsExecutionHandler.SetScheduledRootHashAndSCRs(currHeader.GetRootHash(), make(map[block.Type][]data.TransactionHandler))
+		gasAndFees := scheduled.GasAndFees{
+			AccumulatedFees: big.NewInt(0),
+			DeveloperFees:   big.NewInt(0),
+			GasProvided:     0,
+			GasPenalized:    0,
+			GasRefunded:     0,
+		}
+		boot.scheduledTxsExecutionHandler.SetScheduledRootHasSCRsAndGas(currHeader.GetRootHash(), make(map[block.Type][]data.TransactionHandler), gasAndFees)
```
Comment (raduchis) [inline process/sync/baseSync.go:930]: this is dupplicated code with line 772 from here but also from the baseProcessor RollBackToBlock section and baseBootStrapStorage in the next file in this PR.

**Review by raduchis — CHANGES_REQUESTED** [review]

File: epochStart/bootstrap/startInEpochScheduled.go [hunk 9]
```diff
@@ -218,7 +219,14 @@ func (ses *startInEpochWithScheduledDataSyncer) prepareScheduledSCRs(
...
 	if additionalData != nil {
-		ses.saveScheduledSCRs(scheduledSCRs, header.GetAdditionalData().GetScheduledRootHash(), header.GetPrevHash())
+		gasAndFees := scheduled.GasAndFees{
+			AccumulatedFees: additionalData.GetScheduledAccumulatedFees(),
+			DeveloperFees:   additionalData.GetScheduledDeveloperFees(),
+			GasProvided:     additionalData.GetScheduledGasProvided(),
+			GasPenalized:    additionalData.GetScheduledGasPenalized(),
+			GasRefunded:     additionalData.GetScheduledGasRefunded(),
+		}
+		ses.saveScheduledSCRs(scheduledSCRs, header.GetAdditionalData().GetScheduledRootHash(), header.GetPrevHash(), gasAndFees)
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply epochStart/bootstrap/startInEpochScheduled.go:229]: done

File: epochStart/bootstrap/startInEpochScheduled.go [hunk 10]
```diff
@@ -248,6 +256,7 @@ func (ses *startInEpochWithScheduledDataSyncer) saveScheduledSCRs(
 	scheduledSCRs map[string]data.TransactionHandler,
 	scheduledRootHash []byte,
 	headerHash []byte,
+	gasAndFees scheduled.GasAndFees,
 ) {
 	if len(scheduledRootHash) == 0 {
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply epochStart/bootstrap/startInEpochScheduled.go:261]: changed to verify accumulatedFees as well

[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply epochStart/bootstrap/startInEpochScheduled.go:261]: if the header is created after scheduled SC calls are enabled, then there will be a non empty scheduled root hash.

File: genesis/process/disabled/scheduledTxsExecutionHandler.go [hunk 11]
```diff
@@ -71,6 +86,10 @@ func (steh *ScheduledTxsExecutionHandler) GetScheduledRootHash() []byte {
 func (steh *ScheduledTxsExecutionHandler) SetScheduledRootHash(_ []byte) {
 }
 
+// SetScheduledGasAndFeeMetrics does nothing as it is a disabled component
+func (steh *ScheduledTxsExecutionHandler) SetScheduledGasAndFeeMetrics(_ scheduled.GasAndFees) {
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply genesis/process/disabled/scheduledTxsExecutionHandler.go:90]: done

File: process/sync/baseSync.go (hunk 8 above)
Comment (AdoAdoAdo) [reply process/sync/baseSync.go:930]: Here just the gasAndFees param has been added, with zero values.

File: process/block/baseProcess.go [hunk 12]
```diff
@@ -1582,12 +1594,87 @@ func (bp *baseProcessor) ProcessScheduledBlock(_ data.HeaderHandler, _ data.Body
...
+	if deltaGasPenalized < 0 {
+		log.Error("feeAndGasMetricsDelta",
+			"initial gasPenalized", initialMetrics.GasPenalized,
+			"final gasPenalized", finalMetrics.GasPenalized,
+			"error", process.ErrNegativeValue)
+		return result
+	}
+	deltaGasRefunded := int64(finalMetrics.GasRefunded) - int64(initialMetrics.GasRefunded)
+	if deltaGasRefunded < 0 {
+		log.Error("feeAndGasMetricsDelta",
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply process/block/baseProcess.go:1662]: done

File: epochStart/bootstrap/startInEpochScheduled.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/startInEpochScheduled.go]: ok

File: genesis/process/disabled/scheduledTxsExecutionHandler.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply genesis/process/disabled/scheduledTxsExecutionHandler.go]: ok

File: integrationTests/vm/testInitializer.go [hunk 13]
```diff
@@ -468,6 +480,7 @@ func CreateTxProcessorWithOneSCExecutorMockVM(
 		PenalizedTooMuchGasEnableEpoch: argEnableEpoch.PenalizedTooMuchGasEnableEpoch,
 		BuiltInFunctionsEnableEpoch:    argEnableEpoch.BuiltinEnableEpoch,
 		SCDeployEnableEpoch:            argEnableEpoch.DeployEnableEpoch,
+		ScheduledMiniBlocksEnableEpoch: 1000,
```
[the remark this answers is not in the document]
Comment (raduchis) [reply integrationTests/vm/testInitializer.go:484]: ok

File: epochStart/bootstrap/startInEpochScheduled.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/startInEpochScheduled.go]: ok

File: process/block/baseProcess.go [hunk 14]
```diff
@@ -1199,11 +1202,18 @@ func (bp *baseProcessor) revertAccountState() {
...
 	if err != nil {
-		bp.scheduledTxsExecutionHandler.SetScheduledRootHashAndSCRs(header.GetRootHash(), make(map[block.Type][]data.TransactionHandler))
+		gasAndFees := scheduled.GasAndFees{
+			AccumulatedFees: big.NewInt(0),
+			DeveloperFees:   big.NewInt(0),
+			GasProvided:     0,
+			GasPenalized:    0,
+			GasRefunded:     0,
+		}
+		bp.scheduledTxsExecutionHandler.SetScheduledRootHasSCRsAndGas(header.GetRootHash(), make(map[block.Type][]data.TransactionHandler), gasAndFees)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply process/block/baseProcess.go]: ok

File: process/block/postprocess/feeHandler.go (hunk 1 above)
Comment (raduchis) [reply process/block/postprocess/feeHandler.go]: ok

File: process/block/postprocess/feeHandler.go (hunk 2 above)
Comment (raduchis) [reply process/block/postprocess/feeHandler.go]: ok

File: process/block/postprocess/feeHandler_test.go (hunk 3 above)
Comment (raduchis) [reply process/block/postprocess/feeHandler_test.go]: ok

File: process/block/preprocess/scheduledTxsExecution.go (hunk 4 above)
Comment (raduchis) [reply process/block/preprocess/scheduledTxsExecution.go]: ok

File: process/block/preprocess/scheduledTxsExecution.go (hunk 5 above)
Comment (raduchis) [reply process/block/preprocess/scheduledTxsExecution.go]: ok

File: process/sync/baseSync.go (hunk 8 above)
Comment (raduchis) [reply process/sync/baseSync.go]: ok

File: process/coordinator/process.go (hunk 7 above)
Comment (raduchis) [reply process/coordinator/process.go]: ok

**Review by raduchis — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Pull request merged**